[#1401] Present configuration read errors for the user

fabjan opened the pull request:
Description
========

If the the configuration file(s) found cannot be parsed, show a notice for the user instead of silently moving on to the next config alternative.

Fixes  #1401

fabjan commented:
The notice in VS Code:

![Skärmavbild 2022-11-14 kl  20 18 02](https://user-images.githubusercontent.com/33322/201754855-c1ce4414-cc92-4dd6-923a-4be00adba5e4.png)

Example log message (I don't know why it shows up twice, that happens for me with any version of the language server):

![Skärmavbild 2022-11-14 kl  20 18 15](https://user-images.githubusercontent.com/33322/201754850-2383d4fe-5ce2-4e9b-b67c-d9b933993b7f.png)

fabjan commented on apps/els_core/src/els_config.erl:
```diff
@@ -92,27 +92,37 @@
 
--spec initialize(uri(), map(), map(), boolean()) -> ok.
-initialize(RootUri, Capabilities, InitOptions, ReportMissingConfig) ->
+-spec initialize(uri(), map(), map(), use_els_server | no_els_server) -> ok.
```
My reasoning for changing this parameter was that until I happened upon https://github.com/erlang-ls/erlang_ls/issues/1060 I thought the flag was only about verbosity or the config file missing. But after seeing that issue, it appears the purpose of the flag was to prevent `els_server:send_notification` from being called when the DAP script calls `els_config:initialize/3`.

(Perhaps `els_config:initialize/3` can be removed completely and `els_dap_general_provider.erl` updated to call /4 with `no_els_server`).

Other names for the flag could maybe be `no_notice | show_notice`.

fabjan commented:
Feel free to close this PR if there is too much else going on. I prefer to not be the one closing it since it is not targeting my repository.

robertoaloi commented:
LGTM, but again the CI jobs seem to be expired and could not find a way to re-run them. If you can force-push, we can merge this one.

fabjan commented:
I rebased on master since I was going to force push and get another CI run anyway.

When doing a last minute test locally I found that a bad YAML yields a crash when merging the configs so I added a little `syntax_error` reason for the broken config reporting.

![Skärmavbild 2023-02-11 kl  19 50 00](https://user-images.githubusercontent.com/33322/218276199-1fa54b34-4943-4c2a-bb7d-c43ac17d8006.png)

fabjan commented on apps/els_core/src/els_config.erl:
```diff
@@ -86,33 +86,46 @@
     providers => map()
 }.
 
+-type error_reporting() :: lsp_notification | log.
```
As I read through my changes again after some time away from it I came to the conclusion that `no_els_server` and `use_els_server` were kind of bad names for the variants. I think this is more clear.

fabjan commented on apps/els_core/src/els_config.erl:
```diff
@@ -86,33 +86,46 @@
 %%==============================================================================
 
 -spec initialize(uri(), map(), map()) -> ok.
 initialize(RootUri, Capabilities, InitOptions) ->
-    initialize(RootUri, Capabilities, InitOptions, _ReportMissingConfig = false).
+    %% https://github.com/erlang-ls/erlang_ls/issues/1060
+    initialize(RootUri, Capabilities, InitOptions, _ErrorReporting = log).
```
`initialize/3` could be removed completely unless it is kept for some backwards compatibility I can't see.

robertoaloi replied:
I don't see a reason to keep it, either.

fabjan commented:
@robertoaloi I really didn't mean to spam when I commented the PRs can be closed. I just wanted to make sure they were not going to linger around making a mess!

I will touch them one by one now, so the other two will be "on hold" until this one is merged or closed.

Sorry for scope creeping a bit here after you already reviewed it, but I just changed the names of the error reporting selector variants and a little error handling for configs that are valid YAML but not maps (and thus can't be used at all) which I discovered when picking this back up, rebasing and testing locally.

robertoaloi commented:
> @robertoaloi I really didn't mean to spam when I commented the PRs can be closed.

No problem at all, thanks for the call to action and for the contribution!

robertoaloi commented:
Thanks for the contribution!

fabjan commented:
Looks like the Windows CI failed because chocolatey could not install Erlang 22.3 :(
